rust/ql/test/library-tests/dataflow/sources/test.rs

```diff
@@ -420,10 +420,10 @@ fn test_fs() -> Result<(), Box<dyn std::error::Error>> {
 
     for entry in fs::read_dir("directory")? {
         let e = entry?;
-        let path = e.path(); // $ Alert[rust/summary/taint-sources]
-        let file_name = e.file_name(); // $ Alert[rust/summary/taint-sources]
-        sink(path); // $ hasTaintFlow
-        sink(file_name); // $ hasTaintFlow
+        let path = e.path(); // $ MISSING: Alert[rust/summary/taint-sources]
+        let file_name = e.file_name(); // $ MISSING: Alert[rust/summary/taint-sources]
+        sink(path); // $ MISSING: hasTaintFlow
+        sink(file_name); // $ MISSING: hasTaintFlow
```

Comment:
@hvitved I'm looking to get these results back.  The sources go missing because there's no result for `getCanonicalPath` on lines 423 and 424, I'm guessing that means we haven't inferred the type of `e` / `entry` due to limitations of type inference on `for` loops (line 421).

If you agree that's the problem, do you have an idea what the general case for `inferForLoopExprType` should look like?  I think we need to get the type of the `Iterator` (returned by `fs::read_dir` in this case) and pull it apart for it's `Item` type.

Reply:
You are right, it must be because we cannot currently infer the type of `e`, which means we cannot resolve the `e.path()` call (note that it is not `e.path()` itself that has a canonical path, it is the (missing) target that has a canonical path).

> If you agree that's the problem, do you have an idea what the general case for inferForLoopExprType should look like? I think we need to get the type of the Iterator (returned by fs::read_dir in this case) and pull it apart for it's Item type.

Correct; I will give it a try.